src/macro-ambiguity.md

```diff
@@ -0,0 +1,375 @@
+  - `delimited sequence`: a sequence of token trees with appropriate open- and
+    close-delimiters at the start and end of the sequence.
+  - `empty fragment`: The class of invisible Rust syntax that separates tokens,
+    i.e. whitespace, or (in some lexical contexts), the empty token sequence.
+  - `fragment specifier`: The identifier in a simple NT that specifies which
+    fragment the NT accepts.
+  - `language`: a context-free language.
+
+Example:
+
+```compile_fail
+macro_rules! i_am_an_mbe {
+    (start $foo:expr $($i:ident),* end) => ($foo)
+}
+```
+
+`(start $foo:expr $($i:ident),\* end)` is a matcher. The whole matcher is a
+delimited sequence (with open- and close-delimiters `(` and `)`), and `$foo`
+and `$i` are simple NT's with `expr` and `ident` as their respective fragment
+specifiers.
+
+`$(i:ident),\*` is *also* an NT; it is a complex NT that matches a
+comma-seprated repetition of identifiers. The `,` is the separator token for
+the complex NT; it occurs in between each pair of elements (if any) of the
+matched fragment.
+
+Another example of a complex NT is `$(hi $e:expr ;)+`, which matches any
+fragment of the form `hi <expr>; hi <expr>; ...` where `hi <expr>;` occurs at
+least once. Note that this complex NT does not have a dedicated separator
+token.
+
+(Note that Rust's parser ensures that delimited sequences always occur with
+proper nesting of token tree structure and correct matching of open- and
+close-delimiters.)
+
+We will tend to use the variable "M" to stand for a matcher, variables "t" and
+"u" for arbitrary individual tokens, and the variables "tt" and "uu" for
+arbitrary token trees. (The use of "tt" does present potential ambiguity with
+its additional role as a fragment specifier; but it will be clear from context
+which interpretation is meant.)
+
+"SEP" will range over separator tokens, "OP" over the repetition operators
+`\*`, `+`, and `?`, "OPEN"/"CLOSE" over matching token pairs surrounding a
+delimited sequence (e.g. `[` and `]`).
+
+We also use Greek letters "α" "β" "γ" "δ" to stand for potentially empty
+token-tree sequences. (However, the Greek letter "ε" (epsilon) has a special
+role in the presentation and does not stand for a token-tree sequence.)
+
+  * This Greek letter convention is usually just employed when the presence of
+	a sequence is a technical detail; in particular, when I wish to *emphasize*
+	that we are operating on a sequence of token-trees, I will use the notation
+	"tt ..." for the sequence, not a Greek letter
+
+Note that a matcher is merely a token tree. A "simple NT", as mentioned above,
+is an meta-variable NT; thus it is a non-repetition. For example, `$foo:ty` is
+a simple NT but `$($foo:ty)+` is a complex NT.
+
+Note also that in the context of this formalism, the term "token" generally
+*includes* simple NTs.
+
+Finally, it is useful for the reader to keep in mind that according to the
+definitions of this formalism, no simple NT matches the empty fragment, and
```

Comment:
`vis` matches the empty fragment.

Reply:
I don't think we discussed that in particular, no.

Reply:
Er, sorry, I misread as "implications on the running time" -- in any case, I'm not sure precisely what implications @alercah has in mind. (I haven't read this document.) That said, the behavior of the `vis` matcher is stable, I believe? (And quite useful.)

Reply:
I am not sure what, if any, there are either. But the algorithm pretty explicitly assumes that fragment specifiers can't match empty token sequences, so that makes me think that it ought to have been considered before stabilization. ;)

(I had a related follow-up on the tracking issue about `crate`; I don't know if you saw that? `vis` has possibly too broad of a follow set)

Reply:
> That said, the behavior of the `vis` matcher is stable, I believe? (And quite useful.)

Yup.

Reply:
Okay, I realized that this works fine as long as we just define the algorithm to treat it like `$($v:vis)?`---now that semantics for `?` are defined, possibly-empty matches for metavariables should pose no problem. So I put that in as such.